lib/src/common/managed_stream_subscription.dart

```diff
@@ -1,61 +1,110 @@
 import 'dart:async';
 
-/// An implementation of `StreamSubscription` that provides a [didCancel] future.
+/// An implementation of `StreamSubscription` that provides a [didComplete]
+/// future.
 ///
-/// The [didCancel] future is used to provide an anchor point for removing
-/// internal references to `StreamSubscriptions` if consumers manually cancel the
-/// subscription. This class is not publicly exported.
+/// The [didComplete] future is used to provide an anchor point for removing
+/// internal references to `StreamSubscriptions` if consumers manually cancel
+/// the subscription. This class is not publicly exported.
+///
+/// There are three situations in which [didComplete] will be completed:
+///   1. the managed subscription is canceled
+///   2. the stream is closed
+///   3. the stream sends an error and `cancelOnError` was set to `true`
 class ManagedStreamSubscription<T> implements StreamSubscription<T> {
+  final bool _cancelOnError;
+
   final StreamSubscription<T> _subscription;
 
-  Completer<Null> _didCancel = new Completer();
+  Completer<Null> _didComplete = new Completer();
 
   ManagedStreamSubscription(Stream<T> stream, void onData(T),
-      {Function onError, void onDone(), bool cancelOnError})
-      : _subscription = stream.listen(onData,
-            onError: onError, onDone: onDone, cancelOnError: cancelOnError);
+      {void onError(error, [stackTrace]), void onDone(), bool cancelOnError})
+      : _cancelOnError = cancelOnError ?? false,
+        _subscription = stream.listen(onData, cancelOnError: cancelOnError) {
+    _wrapOnDone(onDone);
+    _wrapOnError(onError);
+  }
+
+  Future<Null> get didComplete => _didComplete.future;
 
   @override
-  Future<E> asFuture<E>([E futureValue]) => _subscription.asFuture(futureValue);
+  bool get isPaused => _subscription.isPaused;
+
+  @override
+  Future<E> asFuture<E>([E futureValue]) {
+    return _subscription.asFuture(futureValue).then((value) {
+      _complete();
+      return value;
+    }).whenComplete(() {
+      _complete();
```

Comment:
I think you can simplify this to:

```dart
return _subscription.asFuture(futureValue).then(_complete);
```

The `Future` returned from `whenComplete` is guaranteed to return the value from the original future (unless an exception is thrown).